packages/fxa-content-server/server/lib/routes/validate-email-domain.js

```diff
@@ -0,0 +1,62 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * This is basically a RPC endpoint for the frontend to resolve a domain to see
+ * if it has MX record, A record, or neither.
+ */
+
+'use strict';
+
+const joi = require('joi');
+const dns = require('dns');
+const resolver = new dns.promises.Resolver();
+const results = ['MX', 'A', 'none'].reduce((accumulator, val) => {
+  accumulator[val] = val;
+  return accumulator;
+}, {});
+const NotFoundErrorCodes = [dns.NODATA, dns.NOTFOUND];
+
+const tryResolveWith = resolveFunc => async domain => {
+  try {
+    const records = await resolveFunc(domain);
+    // We don't do anything with the records
+    return records && records.length;
+  } catch (err) {
+    if (NotFoundErrorCodes.includes(err.code)) {
+      return false;
+    }
+    throw err;
+  }
+};
+
+module.exports = function() {
+  return {
+    method: 'post',
+    path: '/validate-email-domain',
+    validate: {
+      body: {
+        domain: joi
+          .string()
+          .hostname()
+          .required(),
+      },
+    },
+    process: async function(req, res, next) {
+      const { domain } = req.body;
+
+      try {
+        if (await tryResolveWith(resolver.resolveMx.bind(resolver))(domain)) {
```

Comment:
A series of questions about this:

* Are results cached at all? 
* Does this rely on the OS resolver which caches requests, or does it make a fresh request to the upstream DNS server every time? I couldn't figure that out from the node API documentation.
* Do we have any idea how many of these requests will be made assuming common domains are all part of the front-end allow lists that bypass this check? 
* Do requests time out eventually?
* Does this make 2 requests to the DNS resolver if there is no MX record? Can we make a single request to [`resolve`](https://nodejs.org/docs/latest-v10.x/api/dns.html#dns_dnspromises_resolve_hostname_rrtype) and rely upon the array of returned results instead?

Reply:
> Are results cached at all?

From what I gathered in my quick read on the docs of the pertinent Node functions while I was working on this, no caching.

> Does this rely on the OS resolver which caches requests, or does it make a fresh request to the upstream DNS server every time? I couldn't figure that out from the node API documentation.

The Node docs do make its readers hunt around a bit, but the info is there, and the `resolve*` functions "always use the network to perform DNS queries".  `dns.lookup` is the one function that relies on OS facilities.

> Do we have any idea how many of these requests will be made assuming common domains are all part of the front-end allow lists that bypass this check?

Maybe @davismtl have the stats to give an estimate here?  In percentage terms, it is low, but I don't know what it might be in number of requests.

> Do requests time out eventually?

Yes.  There's a timeout error code.  There's no option to set a timeout from the calling code.

> Does this make 2 requests to the DNS resolver if there is no MX record? Can we make a single request to resolve and rely upon the array of returned results instead?

Yes.  And now I cannot recall the line of reasoning that led me to doing it that way.  I'll change it.  👍

Reply:
> Maybe @davismtl have the stats to give an estimate here? In percentage terms, it is low, but I don't know what it might be in number of requests.

Without giving too many numbers about our registrations, here some numbers that might be useful:
- 80% of registrations are with one of the top 20 domains which should be checked on the front-end. So we will check 20% of our registrations at most.
- 5% of our registration emails bounce and this is what we want to reduce.

Reply:
@shane-tomlinson while working on the changes to use a ANY query, I found legitimate reasons to stick with (potentially) two lookups.

From testing the changes, I noticed that Cloudflare (1.1.1.1) does not support ANY, while Google (8.8.8.8) does.  Looks like Cloudflare stopped supporting it in 2015: https://blog.cloudflare.com/deprecating-dns-any-meta-query-type/.  Moreover, https://tools.ietf.org/html/rfc8482 effectivelly officially deprecates ANY.

Reply:
@jrgm, @jbuck - who are our upstream DNS providers?

Reply:
> 80% of registrations are with one of the top 20 domains which should be checked on the front-end. So we will check 20% of our registrations at most.

20% of registrations seems alright to me.

Reply:
> @jrgm, @jbuck - who are our upstream DNS providers?

We don't configure anything special here, so whatever AWS uses and whatever GCP uses. We can run tests on instances if you want confirm that `ANY` requests will work here, but perhaps it is better to make the two requests given that it's deprecated

Reply:
2 requests it is then!